lib/elixir/lib/calendar/datetime.ex

```diff
@@ -889,6 +877,80 @@ defmodule DateTime do
     Calendar.ISO.add_day_fraction_to_iso_days(iso_days, -offset, 86400)
   end
 
+  defp apply_tz_offset_to_naive(%NaiveDateTime{} = naive_datetime, 0, _), do: naive_datetime
```

Comment:
This could conceivably be implemented as an optimization for `NaiveDateTime.add/3` instead, and that could be used inside `from_iso8601/2`. That would give `NaiveDateTime.add/3` very different performance characteristics depending on the size of the interval being added. I'm not sure how you would feel about that. I thought it might be an unwelcome surprise to some users, which is why I went with a private function and only optimized `from_iso8601/2`.